src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.cs

```diff
@@ -945,20 +932,19 @@ public virtual unsafe int IndexOf(string source, string value, int startIndex, i
                 throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_Index);
             }
 
-            // In Everett we used to return -1 for empty string even if startIndex is negative number so we keeping same behavior here.
-            // We return 0 if both source and value are empty strings for Everett compatibility too.
-            if (source.Length == 0)
-            {
-                if (value.Length == 0)
-                {
-                    return 0;
-                }
-                return -1;
-            }
+            // For compat with previous Framework versions, we allow negative values for 'startIndex' only if 'source' is empty.
+            // These will be normalized to startIndex = 0.
 
             if (startIndex < 0)
             {
```

Comment:
maybe it is worth moving this block after the check (options & ValidIndexMaskOffFlags) and in this block we can return either way with handling the case if value length is not 0. something like: 

            if (startIndex < 0)
            {
                if (source.Length == 0)
                {
                     return value.Length == 0 || Compare("", value,...) is true ? 0 : -1;
                }
                else
                {
                    throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_Index);
                }

Also we may think handling he case value.Length == 0 and this.Length > 0 too if we need to.